File: 323.number-of-connected-components-in-an-undirected-graph.py
#
# @lc app=leetcode id=323 lang=python3
#
# [323] Number of Connected Components in an Undirected Graph
#
from typing import *

# @lc code=start
class UF:
    def __init__(self, n):
        self.cnt = n
        self.parents = [i for i in range(n)]
        self.sizes = [1] * n

    def find(self, p):
        while p != self.parents[p]:
            self.parents[p] = self.parents[self.parents[p]]
            p = self.parents[p]
        return p

    def union(self, p, q):
        root_p, root_q = self.find(p), self.find(q)
        if root_p != root_q:
            if self.sizes[root_p] > self.sizes[root_q]:
                self.parents[root_q] = root_p
                self.sizes[root_p] += self.sizes[root_q]
            else:
                self.parents[root_p] = root_q
                self.sizes[root_q] += self.sizes[root_p]
            self.cnt -= 1

class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        uf = UF(n)
        for u, v in edges:
            uf.union(u, v)
        return uf.cnt

# @lc code=end

File: 323.number-of-connected-components-in-an-undirected-graph.cpp
/*
 * @lc app=leetcode id=323 lang=cpp
 *
 * [323] Number of Connected Components in an Undirected Graph
 */
#include <bits/stdc++.h>

using std::vector;

// @lc code=start
class Solution {
public:
    // Plain DFS solution without union find (can also replace DFS with BFS)
    // T: O(V + E)
    // S: O(V + E)
    int countComponents(int n, vector<vector<int>>& edges) {
        vector<vector<int>> adj(n);
        for (const auto& e : edges) {
            adj[e[0]].push_back(e[1]);
            adj[e[1]].push_back(e[0]);
        }

        vector<uint8_t> vis(n, 0);

        int ans = 0;
        for (int i = 0; i < n; ++i) {
            ans += !vis[i] && dfs(i, adj, vis);
        }
        return ans;
    }

    int dfs(int u, const vector<vector<int>>& adj, vector<uint8_t>& vis) {
        for (const int& v : adj[u]) {
            if (!vis[v]) {
                vis[v] = 1;
                dfs(v, adj, vis);
            }
        }
        return 1;
    }
};
// @lc code=end

